File: CPPBlackJack/src/dealer.cpp
#include "dealer.h"

#include <iostream>

#include "card.h"

Dealer::Dealer() {
  m_hand_.reserve(5);
  m_player_num_ = "D";
  m_value_ = 0;
  m_hide_second_ = true;
}

void Dealer::reset_hand() {
  m_hand_.clear();
  m_value_ = 0;
  m_hide_second_ = true;
}

int Dealer::up_card() { return m_hand_[0]->m_value_; }

std::string Dealer::print() {
  auto output = "Player " + m_player_num_ + ": ";
  for (auto i = 0; i < static_cast<int>(m_hand_.size()); i++) {
    if (i == 1 && m_hide_second_) {
      output += "X ";
    } else {
      output += m_hand_[i]->print() + " ";
    }
  }
  for (auto i = static_cast<int>(m_hand_.size()); i < 5; i++) {
    output += "  ";
  }
  output += "\tScore: " + std::to_string(m_value_);
  if (m_value_ > 21) {
    output += " (Bust) ";
  } else {
    output += "        ";
  }
  return output;
}

File: GoBlackJack/dealer.go
package main

import "strings"

import "fmt"

// Dealer class
type Dealer struct {
	MPlayerNum string
	MHand      []*Card
	MValue     int32
	MAces      int32
	MIsSoft    bool
}

// NewDealer constructor
func NewDealer() *Dealer {
	d := new(Dealer)
	d.MPlayerNum = "D"
	return d
}

// ResetHand resets dealer's hand
func (d *Dealer) ResetHand() {
	d.MHand = d.MHand[:0]
	d.MValue = 0
}

// Print prints the dealers number and hand
func (d *Dealer) Print() string {
	var output strings.Builder
	output.WriteString("Player " + fmt.Sprint(d.MPlayerNum) + ": ")
	for _, card := range d.MHand {
		output.WriteString(card.Print() + " ")
	}
	for i := len(d.MHand); i < 5; i++ {
		output.WriteString("  ")
	}
	output.WriteString("\tScore: " + fmt.Sprint(d.MValue))
	if d.MValue > 21 {
		output.WriteString(" (Bust) ")
	} else {
		output.WriteString("        ")
	}
	return output.String()
}

// Evaluate evaluates the player's hand
func (d *Dealer) Evaluate() {
	d.MAces = 0
	d.MValue = 0
	for _, card := range d.MHand {
		d.MValue += card.MValue
		// check for ace
		if card.MIsAce {
			d.MAces++
			d.MIsSoft = true
		}
	}

	for d.MValue > 21 && d.MAces > 0 {
		d.MValue -= 10
		d.MAces--
	}

	if d.MAces == 0 {
		d.MIsSoft = false
	}
}

// UpCard returns first card in dealer's hand
func (d *Dealer) UpCard() int32 {
	return d.MHand[0].MValue
}
